### DISCUSSION TITLE
How to populate component data from PHP-rendered page?

### DISCUSSION THREAD

**Discussion by docdunning** [alpinejs/alpine#4055] [1. Help]
OK, so Alpine's primary target usage is to add interactivity to a server-side-rendered page, rather than producing an SPA.

I previously built a small web site where the Alpine code was fetching all the data from an API, and the HTML was a template for Alpine to complete and activate.  It worked OK, and was a useful exercise.

Now I'm trying the other approach, using PHP to render the data in the HTML, and with the necessary Alpine directives already included.  So - rather than fetching its data from an API, the Alpine code must somehow derive it from the HTML that PHP has created.

In some cases where the data is used in input fields, I can use `x-model:fill` to populate the Alpine data, but that's only for input fields.

I know I could generate JS code in PHP, potentially part of the Alpine.data() block itself, but I'm not altogether happy with that idea.

What's best practice for populating component data in this situation?   Here are a few ideas I've had - I'm interested to know how other Alpine users have approached this.  Fwiw, I think I slightly favour using `x-ref` per approach 2.

```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
  <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.5/dist/cdn.min.js"></script>
</head>
<body>
<div x-data="mydata">

  <!-- Approach 1 using x-init -->
  <p x-init="student.name=$el.innerText">Arthur</p>
  <p x-text="student.name + '!'"></p>

  <!-- Approach 2 using x-ref and init() -->
  <p x-ref="age">32</p>
  <p x-text="student.age + '!'"></p>

  <!-- Approach 3 using custom directive -->
  <p x-copy:student="subject">History</p>
  <p x-text="student.subject + '!'"></p>

  <!-- Approach 4 using hidden input -->
  <input type="hidden" x-model.fill="student.location" value="London">
  <p x-text="student.location + '!'"></p>

</div>
<script>
document.addEventListener('alpine:init', () => {
  Alpine.data('mydata',() => ({
    student: {name: null, age: null, subject: null, location: ''},
    init() {
      this.student.age = this.$refs.age.innerText
    }
  }))
  Alpine.directive('copy', (el, { expression, value }) => {
    if (value) {
      Alpine.$data(el)[value][expression] = el.innerText
    } else {
      Alpine.$data(el)[expression] = el.innerText
    }
  })
})
</script>
</body>
</html>
```

**Comment by ekwoka** [comment]
Just render it as JSON.

In a Django app I work on, a common pattern for these larger blocks is to render a script tag as the first child with JSON in it. the component just grabs and parses the json during initialization.

In your example, why do you need to make that data dynamic though? Why do you need to pull the student name into the context at all?

**Reply by docdunning** [reply]
> Just render it as JSON.

Yes, after thinking it through, I've decided that you're right.  It gives me an easy way to transition from the original code that used the API, because the same data can be picked up from the JSON `<script>.`  If there's a downside, it's that the server and client are tightly bound, but I can't see how that can be avoided anyway. 

Here are the relevant extracts from the code I'm using (modified to use as an example):

```
<?php
$student = get_student(123);
$courses = get_all_courses();
$jdata = [
  'student' => $student,
  'courses' => $courses,
];
?>
<script id="jdata" type="application/json">
<?php echo json_encode($jdata), PHP_EOL; ?>
</script>
```

Then in Alpine:

```
init() {
  const jdata = JSON.parse(document.getElementById('jdata').textContent)
  this.student = jdata.student
  this.courses = jdata.courses
},
```

By the way, I don't really need every piece of server data to be replicated in the Alpine context, so your question "why do you need to make that data dynamic?" is a good one.  I guess it's partly because I'm moving from full-on client-side code, and forgetting that some data is fine just being rendered on the server.  It's only if it's going to change, or if it figures in something else that changes - then I'd need it in the Alpine data context.

**Reply by ekwoka** [reply]
I'd recommend in this example to get used to not using `id` for this, since some components that you use this pattern with could have multiple on a page. I'm partial to `firstElementChild` or at least using a scoped `querySelector` like `this.$root.querySelector(script[type=application/json]`). This can ensure it doesn't try to grab a script somewhere else.

**Reply by docdunning** [reply]
Thanks, @ekwoka, that's useful advice.  It made me reconsider.  

For now I plan to go with `x-ref` because that's very unlikely to pick up the wrong script element, and as a bonus it reduces code clutter.

So

```
<script x-ref="jdata" type="application/json">
<?php echo json_encode($jdata), PHP_EOL; ?>
</script>
```

And

```
init() {
  const jdata = JSON.parse(this.$refs.jdata.textContent)
  ...
```

Edit:
By the way, do you think there's any benefit in using `remove()` to take the script out of the DOM after the data has been picked up?

**Reply by docdunning** [reply]
Duh - I just realised that I can't use `x-ref`, because I'm also using the JSON data to populate parts of an Alpine `store`, and that has no access to  `$refs`.  So I'll have to go back to one of the more general JS selectors.

**Comment by docdunning** [comment]
Largely as a proof of concept, I put together a custom directive to populate Alpine data or store properties with JSON from a script element.  Actually, I think I shall use it in my project, and I'm posting it here in case it's of interest to anyone in a similar situation.

It might seem a bit over the top, but it can provide values to `data()` or `store()` properties; it's self-contained, in that no code is needed in the `data()` or `store()` objects; it can be used more than once on a single script element; and there's no need to find the script element via its HTML `id`.

The list of properties to be copied is given in the modifiers, `x-json.prop1.prop2 ...`  
If I want to copy to a store, I can provide a store name `x-json:storename.prop1.prop2 ...`

Potential downsides - there's no error-checking, and it can only copy to props with the same name.

```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
  <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.5/dist/cdn.min.js"></script>
</head>
<body>
<div x-data="mydata">
  <!-- In reality, the script element contents will be provided by PHP via json_encode($foo); -->
  <script
    type="application/json"
    x-json.student.course
    x-json:global.user
    >
    {
      "student":{"name":"Amy","age":32,"subject":"Art","location":"Glasgow"},
      "course":{"name":"Art","weeks":10,"location":"Perth"},
      "user":"Oscar"
    }
  </script>
  <p x-text="$store.global.user"></p>
  <p><span x-text="student.name"></span> <span x-text="student.location"></span></p>
  <p><span x-text="course.name"></span> <span x-text="course.weeks"></span></p>
</div>
<script>
document.addEventListener('alpine:init', () => {
  Alpine.data('mydata',() => ({
    student: {name: null, age: null, subject: null, location: ''},
    course: {name: null, weeks:null, location:''},
  }))
  Alpine.store('global',{
    user: null,
  })
  Alpine.directive('json', (el, { modifiers, value }) => {
    const jdata = JSON.parse(el.innerText)
    for (const m of modifiers) {
      if (value) {
        Alpine.store(value)[m] = jdata[m]
      } else {
        Alpine.$data(el)[m] = jdata[m]
      }
    }
  })
})
</script>
</body>
</html>
```

**Reply by ekwoka** [reply]
Interesting!

What is the order of operations like?

Would this run before or after the `init` method on the component?

**Reply by docdunning** [reply]
> What is the order of operations like?

I added a few console logs to the example code.

1.  store `init()` function
2.  data `init()` function
3.  `x-init` directive on containing `<div>`
4.  `x-json.student.course`
5.  `x-json:global.user`

Edit: And that gives us a problem, because the two `init()` functions very likely need to use the data provided by the JSON.  Some sort of callback is needed, I think.  Working on it.

**Reply by ekwoka** [reply]
I forget exactly how the tree is initialized, but would the directive `before` attribute help?

**Reply by docdunning** [reply]
> would the directive `before` attribute help?

That's a great suggestion, and I'd not explored it before now.  In this case it doesn't work, though, because it will never be able to get ahead of the `init()` functions.  Actually, in my tests, I couldn't get `.before('init')` to work, either, but no worries.

Here's an updated directive, which calls a function (assuming it exists) after the data is copied across.  It uses the `expression` parameter of the directive to pass the function name.  Btw, thanks to JS for the optional chaining operator!

```
document.addEventListener('alpine:init', () => {
  Alpine.directive('json', (el, { expression, modifiers, value }) => {
    const jdata = JSON.parse(el.innerText)
    for (const m of modifiers) {
      if (value) {
        Alpine.store(value)[m] = jdata[m]
      } else {
        Alpine.$data(el)[m] = jdata[m]
      }
    }
    if (value) {
      Alpine.store(value)[expression]?.()
    } else {
      Alpine.$data(el)[expression]?.()
    }
  })
})
```

Now (e.g.) I can use the directive as follows, so that the `jsonLoaded()` function is invoked after the data is copied over.

```
<script
  type="application/json"
  x-json.student.course="jsonLoaded"
  x-json:global.user
  >
  {
    "student":{"name":"Amy","age":32,"subject":"Art","location":"Glasgow"},
    "course":{"name":"Art","weeks":10,"location":"Perth"},
    "user":"Oscar"
  }
</script>
```

**Reply by docdunning** [reply]
And now with a minor improvement to the directive so that if used without any modifiers, it copies **all** the variables in the script element.

Usage in simplest form:

```
<script type="application/json" x-json>
... { JSON encoded object } ...
</script>

```

```
Alpine.directive('json', (el, { expression, modifiers, value }) => {
  const jdata = JSON.parse(el.innerText)
  if (!modifiers.length) modifiers = Object.keys(jdata)
  for (let m of modifiers) {
    if (value) {
      Alpine.store(value)[m] = jdata[m]
    } else {
      Alpine.$data(el)[m] = jdata[m]
    }
  }
  if (value) {
    Alpine.store(value)[expression]?.()
  } else {
    Alpine.$data(el)[expression]?.()
  }
})

```

**Reply by ekwoka** [reply]
You can just use the `evaluate` helper to just evaluate the expression, so the expression can be `x-json="$store.users"` or whatever.

**Reply by docdunning** [reply]
@ekwoka 

> You can just use the `evaluate` helper to just evaluate the expression, so the expression can be `x-json="$store.users"` or whatever.

Hmm.  You'll have to help me out with that one.  I understand the idea, but I can't see how I'd get it to work.  Is your idea that (in this case) all the JSON in the script element is to be used to populate the store called `users`?

In my use case, the script element will contain an object, with one or more sub-objects.  The contents have been created by PHP.

```
<script type="application/json">
    {
      "student":{"name":"Amy","age":32,"subject":"Art","location":"Glasgow"},
      "course":{"name":"Art","weeks":10,"location":"Perth"},
      "user":"Oscar"
    }
</script>
```

Now I want `student` and `course` to be copied to the current data scope; and `user` to go into a nominated store.  And then I need to be able to execute a nominated function once that's happened, because the `init()` boats have already sailed.

My `x-json` directive does all of that, so I'm OK with it.  Maybe the JSON could be further structured, so that it defines the targets as well, but to be honest I'm happy that the directive code deals with it.

But yes, I'm interested to understand more about using `evaluate` in this context.

**Reply by ekwoka** [reply]
Sorry I meant for the callback.

You can just do `evaluate(expression)` and it will do whatever the callback is, regardless of nesting or store expression or whatever

**Reply by docdunning** [reply]
> You can just do `evaluate(expression)` and it will do whatever the callback is,

Got it - thanks!  That cuts down code and actually makes it more flexible.  The directive is now:

```
Alpine.directive('json', (el, { expression, modifiers, value }, { evaluate }) => {
  const jdata = JSON.parse(el.innerText)
  if (!modifiers.length) modifiers = Object.keys(jdata)
  for (let m of modifiers) {
    if (value) {
      Alpine.store(value)[m] = jdata[m]
    } else {
      Alpine.$data(el)[m] = jdata[m]
    }
  }
  evaluate (expression)
})

```

Using the callback like this invokes a function in the `data()` scope:
`<script type="application/json" x-json.student.course="jsonLoaded">`

Or like this for a function in a store:
`<script type="application/json" x-json:global.user="$store.global.storeLoaded">`

I did wonder about forcing the callback to be in the same store as the one being populated, but it's more flexible this way.

Thanks again, @ekwoka

**Reply by docdunning** [reply]
@ekwoka 

I'm finding that `evaluate()` throws errors if it's given an empty string.  

So if I have `x-json.student.course` (i.e. with no callback), I get:
`Alpine Expression Error: Unexpected token '}'`

No big deal, I just change the directive to have `if (expression) evaluate(expression)`

But is this by design?  Should `evaluate()` itself check for and ignore an empty string?

**Reply by ekwoka** [reply]
It doesn't

**Reply by docdunning** [reply]
OK, I'll just check for an empty string before passing the expression to `evaluate()`.

**Comment by docdunning** [comment]
Given that objects are assigned by reference, I'm wondering about this line of code in the directive.

`Alpine.$data(el)[m] = jdata[m]`

As I understand it, if the item being copied is an object, then the Alpine data property is now a reference to `jdata[m]` rather than a clone.

But `jdata` is temporary - it exists only for the lifetime of the directive function.  Is there a risk that the reference is lost after the directive completes?  I don't have enough experience with this situation in JS.

Is it better to clone the value instead, so as to avoid using a reference? 

`Alpine.$data(el)[m] = structuredClone(jdata[m])`

**Reply by docdunning** [reply]
Answering my own question - there's no issue with using a simple assignment:

`Alpine.$data(el)[m] = jdata[m]`

As I said, my lack of experience in JS led me to ask the question.  I'd thought that the left-hand variable was now a reference to the right-hand variable.  Actually, it's just that both variables now point to the same **value**.

So there's no need to clone the source `jdata` object.

**Reply by ekwoka** [reply]
Yes, JavaScript is garbage collected, and should only have values be lost if nothing references them anyway.
